napalm_eos/eos.py

```diff
@@ -1201,3 +1201,140 @@ def traceroute(self, destination, source='', ttl=0, timeout=0):
+
+        bgp_detail_info = {}
+
+        if neighbor_address:
+            peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[0]['output'])
+
+            # Appending peer_info to final data structure
+            if int(peer_info[0]['remote_as']) not in bgp_detail_info:
+                bgp_detail_info[int(peer_info[0]['remote_as'])] = []
+
+            (bgp_detail_info[int(peer_info[0]['remote_as'])].
+                append(peer_info[0]))
+
+        else:
+            # Using preset template to extract peer info
+            v4_peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[0]['output'])
+
+            v6_peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[1]['output'])
+
+            for peer_info in v4_peer_info:
+                # Appending v4 peer_info to final data structure
+                if int(peer_info['remote_as']) not in bgp_detail_info:
+                    bgp_detail_info[int(peer_info['remote_as'])] = []
+
+                bgp_detail_info[int(peer_info['remote_as'])].append(peer_info)
+
+            for peer_info in v6_peer_info:
+                # Appending v6 peer_info to final data structure
+                if int(peer_info['remote_as']) not in bgp_detail_info:
+                    bgp_detail_info[int(peer_info['remote_as'])] = []
+
+                bgp_detail_info[int(peer_info['remote_as'])].append(peer_info)
+
+        return bgp_detail_info
+
+    def _parse_per_peer_bgp_detail(self, peer_output):
+        """This function parses the raw data per peer and returns a
+        json structure per peer.
+        """
+
+        int_fields = ['accepted_prefix_count', 'local_as', 'remote_as',
+                      'local_port', 'remote_port', 'local_port',
+                      'input_messages', 'output_messages', 'input_updates',
+                      'output_updates', 'messages_queued_out', 'holdtime',
+                      'configured_holdtime', 'keepalive',
+                      'configured_keepalive', 'advertise_prefix_count',
+                      'received_prefix_count']
+
+        peer_details = []
+
+        v4_summary = self.device.run_commands(
+            'show ip bgp summary vrf all', encoding='json')
+
+        v6_summary = self.device.run_commands(
+            'show ipv6 bgp summary vrf all', encoding='json')
+
+        # Using preset template to extract peer info
+        peer_info = (
+            napalm_base.helpers.textfsm_extractor(
+                self, 'bgp_detail', peer_output))
+
+        for item in peer_info:
+            # Issuing command to get number of accepted prefixes for peer
+            try:
+                peer_ver = IPAddress(item['remote_address']).version
+            except Exception:
+                continue
+
+            if peer_ver == 4:
+                accepted_prefix_count = (
+                    v4_summary[0]['vrfs']['default']['peers']
```

Comment:
Yes, I see what you are doing now. 
However, this still can be optimised:

You call ```_parse_per_peer_bgp_detail``` for each address family. As most of the time the requests will be made without having the value for ```neighbor_address``` specified, this method will be called twice.
Inside the  ```_parse_per_peer_bgp_detail``` you make two requests for ```sh ip bgp sum vrf all``` and  ```sh ipv6 bgp sum vrf all```.
Thus a total of 4 requests.

I suggest making the ```sh ip bgp sum vrf all``` requests at the beginning (and again you should execute only one of them, depending on the IP version of ```neighbor_address``` if specified.

That way, your implementation will be twice faster.